porch/pkg/git/package.go

```diff
@@ -18,23 +18,28 @@ import (
 	"context"
 	"crypto/sha1"
 	"encoding/hex"
+	"errors"
 	"fmt"
+	"io"
 	"strings"
 	"time"
 
 	"github.com/GoogleContainerTools/kpt/internal/pkg"
 	kptfile "github.com/GoogleContainerTools/kpt/pkg/api/kptfile/v1"
 	"github.com/GoogleContainerTools/kpt/porch/api/porch/v1alpha1"
 	"github.com/GoogleContainerTools/kpt/porch/pkg/repository"
+	"github.com/go-git/go-git/v5"
 	"github.com/go-git/go-git/v5/plumbing"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
 )
 
 type gitPackageRevision struct {
 	repo          *gitRepository // repo is repo containing the package
 	path          string
 	revision      string
+	lifecycle     v1alpha1.PackageRevisionLifecycle
```

Comment:
Does this need to be a field on the struct? It seems like it is for performance, so do we know that we pick up any changes to the refs in git outside of porch?

Reply:
Yeah, it's for performance. I noticed that looking through all the branches introduced a pretty noticeable lag (maybe 1-2 extra seconds) just for the few package revisions I had, so I imagine it would be much worse as the number of package revisions starts to grow. 

I tested out if having `lifecycle` be a field on the `gitPackageRevision` still picks up changes to refs in git outside of porch, by creating and deleting branches through the Github UI. It did pick up those changes, probably as part of the background refresh. It seemed like it took about a minute to get the updated lifecycles.

However, one issue with that implementation is that we go through all the refs in the repository for _each package revision_ in the repository, but I realized that we only need to do it once per repository. In my most recent commit, I updated this to instead be a field on `gitRepository` so that we don't need to iterate over the refs more than once per repository when listing package revisions. With the new implementation though, we need to make sure we refresh the cache so that we pick up changes outside of porch, so I had to add a couple lines to the cache refresh code for that.